### DISCUSSION TITLE
Handling deploys

### DISCUSSION THREAD

**Discussion by maciejfikus** [spatie/laravel-event-sourcing#267] [Q&A]
Hi guys, 
after playing for a while with this package, it came to the moment when I had to do a production deploy. I wonder how do you manage your deployments in terms of projections/event changes?

Let's say we had added a column to some table and modified projector which manage that, so we have to rebuild the whole table. You deployed a code base, run migrations etc. with help of some deploying tool, what happens next? 

- Do you manually log in into the machine and run given projections? 
- What happens when you have to replay some amount of events after projector modification (or even brand new one)? As for the example, let's assume that the whole process of replaying takes about 10 minutes. In the meantime, your users generated a whole set of new events which means that your projection already out of date as everything that happened during the replay process will not be properly reflected on projected table, am I wrong?

Is maintenance mode the only option to properly introduce new/modified projections?

**Comment by bmbBAMBUS** [comment]
Good question, I am struggling with same issue.

The logic changes required from me to change event properties. I've done it in migration and also ran artisan reapplying projector command from there. I am not sure it's the best solution, but it didn't require changes in deployment process. It also answers your first question.

Maybe using snapshots mechanism would be useful to snap aggregate state before deploy changes, and after that use events no more, use snapshot state instead. Just thinking loud.

Any another solutions/thoughts?

**Comment by maciejfikus** [comment]
Hi @bmbBAMBUS 
Migrations won't suffice I suppose. In fact, I could run projector commands from there but it doesn't seem right. I can image situations where I would need to run migrations without running projections. We could work out some similar concept dedicated for projector replayers, however my second issue should be resolved first in order to came out with proper solution here.

Using snapshots definitely would help with performance but there is still a potential gap in scenario I describe in my first post.
